internal/app/referenceserver/impl.go

```diff
@@ -192,6 +192,12 @@ func (s *conformanceServer) BidiStream(
 			responseDefinition = req.ResponseDefinition
 			fullDuplex = req.FullDuplex
 			firstRecv = false
+
+			// If a response definition was provided, add the headers and trailers
+			if responseDefinition != nil {
+				internal.AddHeaders(responseDefinition.ResponseHeaders, stream.ResponseHeader())
+				internal.AddHeaders(responseDefinition.ResponseTrailers, stream.ResponseTrailer())
+			}
```

Comment:
I would move this to the very top. Setting the headers is a no-op if they've already been sent (which happens with the first call to `stream.Send`), so we want to make sure they are set as the first thing.

Reply:
This is actually only run once when we receive the first message (and before any sends). I _thiink_ it should be OK where it is based on what you mentioned, but let me know if I'm missing something.

Reply:
I wasn't saying it's in the wrong place. I was suggesting moving to the top because that's a more intuitive place for it IMO. Right now, you have to look closely at the code to decide if this is a bug or not. Whereas if it were at the top, it would be obviously correct.

Actually, I just spotted a problem, so it does actually _need_ to move: if the request stream has _zero_ messages, this will never get called.

Reply:
Oof, never mind, I'm an idiot. 🤦 

You have to read the first request in order to know what headers/trailers to set. My bad.

Reply:
![Homer hiding](https://media3.giphy.com/media/jUwpNzg9IcyrK/giphy.gif?cid=ecf05e471s1uyl4jyws9lh7pfo233ywgwicg204tpwjtj2kq&ep=v1_gifs_search&rid=giphy.gif&ct=g)